File: VulkanEngine/Core/Entities/Light.cs
using VulkanEngine.Core.Math;

namespace VulkanEngine.Core.Entities;

public enum LightType
{
    Directional = 0,
    Point = 1,
    Ambient = 2
}

public class Light
{
    public LightType Type { get; set; }
    public Vector3 Position { get; set; }
    public Vector3 Direction { get; set; }
    public Color Color { get; set; }
    public float Intensity { get; set; }

    public Light(LightType type, Color? color = null, float intensity = 1.0f)
    {
        Type = type;
        Color = color ?? Math.Color.White;
        Intensity = intensity;
        Position = Vector3.Zero;
        Direction = new Vector3(0, -1, 0);
    }

    public static Light CreateDirectional(Vector3 direction, Color? color = null, float intensity = 1.0f)
    {
        return new Light(LightType.Directional, color, intensity)
        {
            Direction = direction.Normalized
        };
    }

    public static Light CreatePoint(Vector3 position, Color? color = null, float intensity = 1.0f)
    {
        return new Light(LightType.Point, color, intensity)
        {
            Position = position
        };
    }

    public static Light CreateAmbient(Color? color = null, float intensity = 0.3f)
    {
        return new Light(LightType.Ambient, color, intensity);
    }
}

File: web/src/Core/Light.ts
import { Vector3 } from './Math';
import { Color } from './Math';

export enum LightType {
    DIRECTIONAL = 0,
    POINT = 1,
    AMBIENT = 2
}

export class Light {
    public type: LightType;
    public position: Vector3;
    public direction: Vector3;
    public color: Color;
    public intensity: number;

    constructor(
        type: LightType = LightType.DIRECTIONAL,
        position: Vector3 = Vector3.zero(),
        direction: Vector3 = new Vector3(0, -1, 0),
        color: Color = Color.white(),
        intensity: number = 1.0
    ) {
        this.type = type;
        this.position = position;
        this.direction = direction.normalize();
        this.color = color;
        this.intensity = intensity;
    }

    static createDirectional(direction: Vector3, color: Color = Color.white(), intensity: number = 1.0): Light {
        return new Light(LightType.DIRECTIONAL, Vector3.zero(), direction, color, intensity);
    }

    static createPoint(position: Vector3, color: Color = Color.white(), intensity: number = 1.0): Light {
        return new Light(LightType.POINT, position, Vector3.zero(), color, intensity);
    }

    static createAmbient(color: Color = Color.white(), intensity: number = 0.15): Light {
        return new Light(LightType.AMBIENT, Vector3.zero(), Vector3.zero(), color, intensity);
    }
}


